### ISSUE TITLE
Is it possible to cancel an io device await?

### ISSUE THREAD

**Issue by dabeibao** [danvratil/qcoro#53]
For example, send a command via serial port; await its response but need to cancel it due to timeout.

**Comment by danvratil** [comment]
Technically, it's possible to add a timeout argument that would cause the coroutine to "self-cancel". There's no way to cancel a coroutine from the outside, though if that's what you are asking.

`QIODevice::read()` doesn't accept a timeout argument and I'd like the QCoro API to stay as close to the Qt API as possible, but I could add a wrapper for `QIODevice::waitForReadyRead()`, so you would send the command and then do:
```
const bool ready = co_await qCoro(myDevice).waitForReadyRead(timeout);
if (ready) {
   const auto data = co_await qCoro(mDevice).readLine();
} else {
  // handle timeout
}
```

Would that work for you?

**Comment by dabeibao** [comment]
Thanks! This should work for me.

But I prefer if there is a common solution. Timeout is only one of the scenario. Another possible scenario is user click a button the cancel the command.
Is it possible that the API provides a context parameter so user can use the context to cancel the await?

**Comment by danvratil** [comment]
You can't easily "cancel" a coroutine. In this case, coroutines are functions like any other, and you can't just cancel a running function. Even if the coroutine is currently suspended, destroying the associated state before the coroutine has finished would lead to memory leaks (at best).

Some operations are not cancellable, or cancellation is implementation defined, e.g. if you want to cancel a command you sent to a remote server do you just stop waiting for the response, or do you need to send something to the server to cancel processing the request etc.?

If you need cancellation for some of your operations, I'd suggest following:

* you pass some kind of a cancellation token to your coroutine to allow cancelling the underlying operation
* `await` the coroutine
* if user clicks a button, you cancel the underlying operation using the cancellation token
* when the `await`ed coroutine finishes, you check the state of the cancellation token to know whether the underlying operation has finished or whether it was canceled

**Comment by dabeibao** [comment]
> * you pass some kind of a cancellation token to your coroutine to allow cancelling the underlying operation
> 
> * `await` the coroutine
> 
> * if user clicks a button, you cancel the underlying operation using the cancellation token
> 
> * when the `await`ed coroutine finishes, you check the state of the cancellation token to know whether the underlying operation has finished or whether it was canceled

This is exactly what I want. The scenario may be
```c++
// coroutine 1
bool ready = await co_await qCoro(myDevice).waitForReadyRead(main_cancellation_token);
if (!ready) {
     // some way to get the error status
} else {
    // handle IO
}
```
```c++
// main
void on_user_click()
{
    main_cancellation_taken.trigger();
}
void on_command_timeont()
{
    main_cancellation_token.trigger();
}
```

So, if the new API `qCoro(myDevice).waitForReadyRead` can accept a cancellation token, we can use it for both timeout and user trigger events.
Per my understanding, it is not easy for users of qCoro to provide the cancellation themself. We need some support from qCoro library

**Comment by danvratil** [comment]
Sorry, I think cancellation tokens are out-of-scope for the current QCoro API, as I'm trying to stick as close to the original Qt API as possible (as QCoro is supposed to be just a wrapper for the Qt API). What I envisioned was that you would pass the cancellation token to your IO device, maybe by subclassing the Qt type you are using, and handle the cancellation of those operations yourself.

As a side-note, I've been prototyping a QCoroIO library with the idea of creating a fully asynchronous IO library for Qt (since I find the current Qt IO API rather frustrating). While I did not think of cancellation tokens initially, I think I'll include them in the API. I don't have any useful code to publish yet, though, so no promises :)
